# Exercise 02 -  OLAP Cubes - CUBE

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [1]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

### Connect to the local database where Pagila is loaded

In [2]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# CUBE 
- Group by CUBE (dim1, dim2, ..) , produces all combinations of different lenghts in one go.
- This view could be materialized in a view and queried which would save lots repetitive aggregations

TODO: Write a query that calculates the various levels of aggregation done in the grouping sets exercise (total, by month, by country, by month & country) using the CUBE function. Your output should match the table below.


In [4]:
%%time
%%sql
SELECT dimDate.month, dimCustomer.country, sum(sales_amount) as revenue
FROM factSales
JOIN dimDate ON (dimDate.date_key =  factSales.date_key)
JOIN dimCustomer ON (dimCustomer.customer_key = factSales.customer_key)
GROUP BY cube(dimDate.month, dimCustomer.country);

 * postgresql://student:***@127.0.0.1:5432/pagila_star
598 rows affected.
CPU times: user 5.03 ms, sys: 180 µs, total: 5.21 ms
Wall time: 40.7 ms


[(1, 'Algeria', Decimal('33.92')),
 (1, 'Angola', Decimal('27.93')),
 (1, 'Anguilla', Decimal('6.97')),
 (1, 'Argentina', Decimal('81.78')),
 (1, 'Austria', Decimal('22.92')),
 (1, 'Azerbaijan', Decimal('20.97')),
 (1, 'Bahrain', Decimal('3.99')),
 (1, 'Bangladesh', Decimal('19.95')),
 (1, 'Belarus', Decimal('5.98')),
 (1, 'Bolivia', Decimal('4.97')),
 (1, 'Brazil', Decimal('210.50')),
 (1, 'Brunei', Decimal('5.99')),
 (1, 'Bulgaria', Decimal('9.98')),
 (1, 'Cambodia', Decimal('11.96')),
 (1, 'Cameroon', Decimal('13.97')),
 (1, 'Canada', Decimal('33.93')),
 (1, 'Chad', Decimal('12.96')),
 (1, 'Chile', Decimal('24.95')),
 (1, 'China', Decimal('455.91')),
 (1, 'Colombia', Decimal('47.87')),
 (1, 'Congo, The Democratic Republic of the', Decimal('43.92')),
 (1, 'Czech Republic', Decimal('0.99')),
 (1, 'Dominican Republic', Decimal('13.97')),
 (1, 'Ecuador', Decimal('20.95')),
 (1, 'Egypt', Decimal('34.91')),
 (1, 'Estonia', Decimal('9.98')),
 (1, 'Faroe Islands', Decimal('6.99')),
 (1, 'Finland', Decimal('3.99')),
 (1, 'France', Decimal('39.92')),
 (1, 'French Guiana', Decimal('5.98')),
 (1, 'French Polynesia', Decimal('29.94')),
 (1, 'Gambia', Decimal('14.97')),
 (1, 'Germany', Decimal('89.80')),
 (1, 'Greece', Decimal('27.92')),
 (1, 'Greenland', Decimal('17.94')),
 (1, 'Holy See (Vatican City State)', Decimal('5.98')),
 (1, 'India', Decimal('429.89')),
 (1, 'Indonesia', Decimal('135.69')),
 (1, 'Iran', Decimal('55.84')),
 (1, 'Israel', Decimal('39.89')),
 (1, 'Italy', Decimal('51.89')),
 (1, 'Japan', Decimal('288.35')),
 (1, 'Kazakstan', Decimal('5.97')),
 (1, 'Kenya', Decimal('7.97')),
 (1, 'Kuwait', Decimal('4.99')),
 (1, 'Latvia', Decimal('12.97')),
 (1, 'Liechtenstein', Decimal('14.98')),
 (1, 'Lithuania', Decimal('9.98')),
 (1, 'Madagascar', Decimal('0.99')),
 (1, 'Malawi', Decimal('1.99')),
 (1, 'Malaysia', Decimal('30.94')),
 (1, 'Mexico', Decimal('263.37')),
 (1, 'Morocco', Decimal('32.94')),
 (1, 'Mozambique', Decimal('28.95')),
 (1, 'Myanmar', Decimal('14.95')),
 (1, 'Nauru', Decimal('4.99')),
 (1, 'Nepal', Decimal('15.97')),
 (1, 'Netherlands', Decimal('25.94')),
 (1, 'New Zealand', Decimal('6.99')),
 (1, 'Nigeria', Decimal('134.71')),
 (1, 'North Korea', Decimal('5.98')),
 (1, 'Oman', Decimal('25.94')),
 (1, 'Pakistan', Decimal('51.88')),
 (1, 'Paraguay', Decimal('1.98')),
 (1, 'Peru', Decimal('40.91')),
 (1, 'Philippines', Decimal('161.62')),
 (1, 'Poland', Decimal('54.89')),
 (1, 'Puerto Rico', Decimal('15.95')),
 (1, 'Romania', Decimal('16.96')),
 (1, 'Runion', Decimal('4.99')),
 (1, 'Russian Federation', Decimal('225.40')),
 (1, 'Saint Vincent and the Grenadines', Decimal('15.97')),
 (1, 'Saudi Arabia', Decimal('65.87')),
 (1, 'Senegal', Decimal('4.99')),
 (1, 'Slovakia', Decimal('8.97')),
 (1, 'South Africa', Decimal('99.76')),
 (1, 'South Korea', Decimal('46.88')),
 (1, 'Spain', Decimal('56.85')),
 (1, 'Sri Lanka', Decimal('12.97')),
 (1, 'Sudan', Decimal('24.95')),
 (1, 'Sweden', Decimal('4.99')),
 (1, 'Switzerland', Decimal('3.98')),
 (1, 'Taiwan', Decimal('55.84')),
 (1, 'Tanzania', Decimal('18.95')),
 (1, 'Thailand', Decimal('17.96')),
 (1, 'Tonga', Decimal('8.98')),
 (1, 'Turkey', Decimal('120.73')),
 (1, 'Turkmenistan', Decimal('9.99')),
 (1, 'Tuvalu', Decimal('26.96')),
 (1, 'Ukraine', Decimal('54.89')),
 (1, 'United Arab Emirates', Decimal('12.95')),
 (1, 'United Kingdom', Decimal('73.84')),
 (1, 'United States', Decimal('309.30')),
 (1, 'Venezuela', Decimal('24.92')),
 (1, 'Vietnam', Decimal('33.91')),
 (1, 'Virgin Islands, U.S.', Decimal('0.99')),
 (1, 'Yemen', Decimal('15.94')),
 (1, 'Yugoslavia', Decimal('25.94')),
 (1, 'Zambia', Decimal('12.97')),
 (1, None, Decimal('4824.43')),
 (2, 'Afghanistan', Decimal('4.97')),
 (2, 'Algeria', Decimal('60.88')),
 (2, 'American Samoa', Decimal('23.95')),
 (2, 'Angola', Decimal('35.92')),
 (2, 'Anguilla', Decimal('16.95')),
 (2, 'Argentina', Decimal('219.49')),
 (2, 'Armenia', Decimal('12.98')),
 (2, 'Austria', Decimal('39.89')),
 (2, 'Azerbaijan', Decimal('25.

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>

## Revenue Total, by Month, by Country, by Month & Country All in one shot, NAIVE way
The naive way to create the same table as above is to write several queries and UNION them together. Grouping sets and cubes produce queries that are shorter to write, easier to read, and more performant. Run the naive query below and compare the time it takes to run to the time it takes the cube query to run.

In [5]:
%%time
%%sql
SELECT  NULL as month, NULL as country, sum(sales_amount) as revenue
FROM factSales
    UNION all 
SELECT NULL, dimStore.country,sum(sales_amount) as revenue
FROM factSales
JOIN dimStore on (dimStore.store_key = factSales.store_key)
GROUP by  dimStore.country
    UNION all 
SELECT cast(dimDate.month as text) , NULL, sum(sales_amount) as revenue
FROM factSales
JOIN dimDate on (dimDate.date_key = factSales.date_key)
GROUP by dimDate.month
    UNION all
SELECT cast(dimDate.month as text),dimStore.country,sum(sales_amount) as revenue
FROM factSales
JOIN dimDate     on (dimDate.date_key         = factSales.date_key)
JOIN dimStore on (dimStore.store_key = factSales.store_key)
GROUP by (dimDate.month, dimStore.country)

 * postgresql://student:***@127.0.0.1:5432/pagila_star
18 rows affected.
CPU times: user 7.92 ms, sys: 0 ns, total: 7.92 ms
Wall time: 43.2 ms


[(None, None, Decimal('67416.51')),
 (None, 'Australia', Decimal('33726.77')),
 (None, 'Canada', Decimal('33689.74')),
 ('1', None, Decimal('4824.43')),
 ('3', None, Decimal('23886.56')),
 ('4', None, Decimal('28559.46')),
 ('2', None, Decimal('9631.88')),
 ('5', None, Decimal('514.18')),
 ('4', 'Canada', Decimal('14423.39')),
 ('3', 'Australia', Decimal('12060.33')),
 ('5', 'Canada', Decimal('243.10')),
 ('1', 'Australia', Decimal('2364.19')),
 ('2', 'Canada', Decimal('4736.78')),
 ('2', 'Australia', Decimal('4895.10')),
 ('1', 'Canada', Decimal('2460.24')),
 ('5', 'Australia', Decimal('271.08')),
 ('3', 'Canada', Decimal('11826.23')),
 ('4', 'Australia', Decimal('14136.07'))]